Arrowgene.Ddon.GameServer/Characters/CharacterManager.cs

```diff
@@ -0,0 +1,85 @@
+using Arrowgene.Ddon.Database;
+using Arrowgene.Ddon.Server.Network;
+using Arrowgene.Ddon.Shared.Entity.PacketStructure;
+using Arrowgene.Ddon.Shared.Entity.Structure;
+using Arrowgene.Ddon.Shared.Model;
+using System;
+using System.Collections.Generic;
+using System.Data.Entity;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Arrowgene.Ddon.GameServer.Characters
+{
+    public class CharacterManager
+    {
+        public static readonly uint BASE_HEALTH = 760U;
+        public static readonly uint BASE_STAMINA = 750U;
+
+        private readonly IDatabase _Database;
+
+        public CharacterManager(IDatabase Database) 
+        {
+            _Database = Database;
+        }
+
+        public void UpdateCharacterExtendedParams(Character Character)
+        {
+            var ExtendedParams = Character.ExtendedParams;
+
+            /**
+             * There are two physical attack traits and two magic attack traits in
+             * the stats menu. This corresponds with the attack of the weapon and
+             * the iniate attack of the character on the currently selected job.
+             * Similar distinction made with magic. The Gain* stats are extra stats
+             * on top of the iniate stats. These come from armors and BO/HO trees.
+             */
+            Character.StatusInfo.GainAttack = ExtendedParams.Attack;
+            Character.StatusInfo.GainDefense = ExtendedParams.Defence;
+            Character.StatusInfo.GainMagicAttack = ExtendedParams.MagicAttack;
+            Character.StatusInfo.GainMagicDefense = ExtendedParams.MagicDefence;
+            Character.StatusInfo.GainStamina = ExtendedParams.StaminaMax;
+            Character.StatusInfo.GainHP = ExtendedParams.HpMax;
+
+            // These values reflect in the UI what the health bar displays
+            // TODO: Remove this when C2SLobbyLobbyDataMsgReq parsing is supported
+            // TODO: Server alone is not able to account for HP added by gear
+            // TODO: So we need to "trust" the value sent to the server is correct
+            Character.StatusInfo.HP = CharacterManager.BASE_HEALTH + ExtendedParams.HpMax;
+            Character.StatusInfo.WhiteHP = CharacterManager.BASE_HEALTH + ExtendedParams.HpMax;
+        }
+
+        public void UpdateCharacterExtendedParamsNtc(GameClient Client, Character Character)
+        {
+            UpdateCharacterExtendedParams(Character);
+            NotifyClientOfCharacterStatus(Client, Character);
+        }
+
+        private void NotifyClientOfCharacterStatus(GameClient Client, Character Character)
+        {
+            S2CContextGetLobbyPlayerContextNtc ntc1 = new S2CContextGetLobbyPlayerContextNtc();
+            GameStructure.S2CContextGetLobbyPlayerContextNtc(ntc1, Character);
+            Client.Send(ntc1);
+
+            S2CCharacterGetCharacterStatusNtc ntc2 = new S2CCharacterGetCharacterStatusNtc
```

Comment:
This packet seems like it should be able to update the character rings, etc. but it doesn't seem to. I can't actually find how this packet definition was actually sourced though. I suspect there may be something important missing or I am using it improperly.

Reply:
Have you tried finding S2C_EXTEND_EQUIP_SLOT_NTC?

Reply:
The only packet I could find which had any meaningful impact on jewelry was `CDataCharacterInfo`. I forget if I tried this equip one, but I can try it again. The main issue is that -- suppose you log in and buy 2 jewelry slots. If you open up the equipment menu and then look, you can see you have 2 slots available. When you try to equip the jewelry though it will show 0/0 and not let you pick anything from the inventories.

![image](https://github.com/sebastian-heinz/Arrowgene.DragonsDogmaOnline/assets/3587094/d5c7cc24-48c0-4e57-b508-365397ca5491)

Logging out and back in, we are able to set it in `CDataCharacterInfo` and then it appears properly and you can equip them. My guess is there must be some equivalent way to update the `CDataCharacterInfo` besides at login.

Reply:
To edit the starting jewelry slots edit line 96 of CreateCharacterHandler.

Reply:
Cool, thanks for the help offline. Using the unidentified NTC packet, was able to update the jewelry properly so this appears to be solved now.